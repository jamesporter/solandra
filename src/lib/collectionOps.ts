export function pairWise<T>(items: T[]): [T, T][] {
  if (items.length < 2) return []
  const res: [T, T][] = []
  for (let i = 0; i < items.length - 1; i++) {
    res.push([items[i], items[i + 1]])
  }
  return res
}

export function tripleWise<T>(items: T[], looped?: boolean): [T, T, T][] {
  if (items.length < 3) return []
  const res: [T, T, T][] = []

  if (looped) res.push([items[items.length - 2], items[0], items[1]])
  for (let i = 0; i < items.length - 2; i++) {
    res.push([items[i], items[i + 1], items[i + 2]])
  }
  if (looped)
    res.push([items[items.length - 2], items[items.length - 1], items[1]])
  return res
}

export function zip2<T, S>(items: T[], other: S[]): [T, S][] {
  const res: [T, S][] = []
  for (let i = 0; i < items.length && i < other.length; i++) {
    res.push([items[i], other[i]])
  }
  return res
}

export function sum(numbers: number[]): number {
  return numbers.reduce((a, b) => a + b, 0)
}

export function arrayOf<T>(n: number, init: () => T): T[] {
  return Array.from({ length: n }, (_) => init())
}

const collectionOps = {
  pairWise,
  tripleWise,
  zip2,
  sum,
  arrayOf,
}

export default collectionOps
