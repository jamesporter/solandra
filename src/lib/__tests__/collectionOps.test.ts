import { describe, it, expect } from 'vitest';
import { pairWise, tripleWise } from "../collectionOps";

describe("Pairwise", () => {
  it("should be able to do pairwise on simple array", () => {
    expect(pairWise([1, 2, 3, 4, 5])).toMatchSnapshot();
  });

  it("should give empty array when not long enough", () => {
    expect(pairWise([1])).toEqual([]);
  });
});

describe("Triplewise", () => {
  it("should be able to do triplewise on simple array", () => {
    expect(tripleWise([1, 2, 3, 4, 5])).toMatchSnapshot();
  });

  it("should give empty array when not long enough", () => {
    expect(tripleWise([1, 2])).toEqual([]);
  });
});

// TODO add back into main tests
// describe("Shuffle", () => {
//   it("should have same elements", () => {
//     const samples = [[1, 2, 3, 4, 6, 9], [1], [1, 2, 3, 4, 6, 9, 12, 21, -2]];

//     for (let sample of samples) {
//       expect([...new Set(shuffle(sample))].sort()).toEqual(
//         [...new Set(sample)].sort()
//       );
//     }
//   });
// });
