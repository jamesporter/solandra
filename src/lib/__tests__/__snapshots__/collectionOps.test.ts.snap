// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pairwise > should be able to do pairwise on simple array 1`] = `
[
  [
    1,
    2,
  ],
  [
    2,
    3,
  ],
  [
    3,
    4,
  ],
  [
    4,
    5,
  ],
]
`;

exports[`Pairwise should be able to do pairwise on simple array 1`] = `
[
  [
    1,
    2,
  ],
  [
    2,
    3,
  ],
  [
    3,
    4,
  ],
  [
    4,
    5,
  ],
]
`;

exports[`Triplewise > should be able to do triplewise on simple array 1`] = `
[
  [
    1,
    2,
    3,
  ],
  [
    2,
    3,
    4,
  ],
  [
    3,
    4,
    5,
  ],
]
`;

exports[`Triplewise should be able to do triplewise on simple array 1`] = `
[
  [
    1,
    2,
    3,
  ],
  [
    2,
    3,
    4,
  ],
  [
    3,
    4,
    5,
  ],
]
`;
